src/Microsoft.ML.Transforms/SvmLight/SvmLightLoader.cs

```diff
@@ -0,0 +1,851 @@
+                    return false;
+                }
+
+                int index = memory.Span.LastIndexOf(separator);
+                if (index == -1)
+                {
+                    left = memory;
+                    right = default;
+                    return false;
+                }
+
+                left = memory.Slice(0, index);
+                right = memory.Slice(index + 1, memory.Length - index - 1);
+                return true;
+            }
+        }
+
+        // This class is used by the CustomMappingTransformer that does the initial parsing of the input.
+        // The features are mapped to two fields: a vector of floats for the feature values, and a vector
+        // of ReadOnlyMemory<char> for the feature indices/names.
+        private sealed class IntermediateInput
+        {
+            public float Label;
+            public float Weight;
+            public VBuffer<ReadOnlyMemory<char>> FeatureKeys;
+            public VBuffer<float> FeatureValues;
+            public ReadOnlyMemory<char> Comment;
+            [KeyType(ulong.MaxValue - 1)]
+            public ulong GroupId;
+        }
+
+        /// <summary>
+        /// This class is used for mapping the vector of <see cref="ReadOnlyMemory{T}"/> of <see cref="char"/> field in an <see cref="IntermediateInput"/> object
+        /// to a numeric vector of indices. This class is used in case the indices in the input file are numeric. For the case where the input file contains
+        /// feature names, instead of a <see cref="CustomMappingTransformer{TSrc, TDst}"/>, we use a <see cref="ValueToKeyMappingTransformer"/>.
+        /// </summary>
+        private sealed class Indices
+        {
+            public VBuffer<int> FeatureKeys;
+
+            public static void ParseIndicesToOneBased(IntermediateInput input, Indices output)
+            {
+                var editor = VBufferEditor.Create(ref output.FeatureKeys, input.FeatureKeys.Length);
+                var inputValues = input.FeatureKeys.GetValues();
+                for (int i = 0; i < inputValues.Length; i++)
+                {
+                    if (Conversions.Instance.TryParse(in inputValues[i], out int index) && index > 0)
+                        editor.Values[i] = index - 1;
+                    else
+                        editor.Values[i] = -1;
+                }
+                output.FeatureKeys = editor.Commit();
+            }
+
+            public static void ParseIndicesToZeroBased(IntermediateInput input, Indices output)
+            {
+                var editor = VBufferEditor.Create(ref output.FeatureKeys, input.FeatureKeys.Length);
+                var inputValues = input.FeatureKeys.GetValues();
+                for (int i = 0; i < inputValues.Length; i++)
+                {
+                    if (Conversions.Instance.TryParse(in inputValues[i], out int index) && index >= 0)
+                        editor.Values[i] = index;
+                    else
```

Comment:
What is the desired behavior if the data is bad? E.g. if the data has duplicate keys? (Maybe throw an exception if the other data loaders throw an exception when the data is loaded?)

Reply:
I added support for keeping track which indices exist in each row (see the `OutputMapper` class in line 359). Currently, I am not throwing in case of duplicate keys, just using the value in the first appearance.
Do you think I should throw instead?

---
In reply to: [332209060](https://github.com/dotnet/machinelearning/pull/4190#discussion_r332209060) [](ancestors = 332209060)

Reply:
For the both the questions above, I would defer to precedence. If other data loaders throw when bad data is encountered, we should throw here as well.

---
In reply to: [333656937](https://github.com/dotnet/machinelearning/pull/4190#discussion_r333656937) [](ancestors = 333656937,332209060)